File: leetcode/problems/problem-specific/647.palindromic-substrings/647.palindromic-substrings.go
package main

// * Two Pointer Solution | O(n^2) Time | O(1) Space

func countSubstrings(s string) int {
	res := 0
	for i := range s {
		// * Handle Odd length
		countSubstringsHelper(s, i, i, &res)

		// * Handle Even length
		countSubstringsHelper(s, i, i+1, &res)
	}

	return res
}

func countSubstringsHelper(s string, left, right int, res *int) {
	sSize := len(s)
	for left >= 0 && right < sSize && s[left] == s[right] {
		*res++
		left--
		right++
	}
}

File: leetcode/problems/sessions/2023-05-01_2023-09-16/647.palindromic-substrings.py
#
# @lc app=leetcode id=647 lang=python3
#
# [647] Palindromic Substrings
#

# @lc code=start

# * Two Pointers Solution | O(n^2) Time | O(1) Space
# * n -> The length of s string


class Solution:
    def countSubstrings(self, s: str) -> int:
        res = 0

        def helper(left, right):
            nonlocal res

            while left >= 0 and right < len(s) and s[left] == s[right]:
                res += 1
                left -= 1
                right += 1

        for i in range(len(s)):
            helper(i, i)
            helper(i, i + 1)

        return res


# @lc code=end
